data/blog/2025-02-01-flow-control.mdx

```diff
@@ -0,0 +1,130 @@
+
+This blog is divided into four sections for easier reading:  
+- [Motivation](#motivation): Why we developed Flow-Control.  
+- [How It Works](#how-it-works): Understanding rate limiting and parallelism.  
+- [How to Use It](#how-to-use-it): Practical examples of implementing Flow-Control in QStash and Workflows.  
+- [What Happened to Queues](#what-happened-to-queues): Changes from the previous queue-based system and the benefits of the new approach.  
+
+## Motivation  
+
+Our community drives our development. We’ve heard your feedback on two key points:  
+
+- **Workflows**: Our [Workflows](https://upstash.com/docs/workflow/agents/overview) product is gaining traction. However, it launched without built-in rate or parallelism limits. The existing parallelism control via Queues was unsuitable for Workflows.  
+- **Queue Limitations**: Many users relied on Queue Parallelism to prevent bursts on their endpoints. However, queues had per-plan limits due to memory/CPU allocation constraints. The new design removes these restrictions, allowing you to configure limits based on your application’s needs without worrying about plan limits.  
+
+To solve these issues, we developed Flow-Control, which works for both Workflows and QStash.  
+
+## How It Works  
+
+- **Rate Limit**: This defines the maximum number of calls per second. Calls within the same `FlowControl` key contribute to the rate count. Instead of rejecting excess calls, QStash queues them for execution in later seconds, respecting the limit.  
+- **Parallelism Limit**: This controls the number of concurrent executions. Unlike rate limiting, execution duration matters. At no time will there be more than the specified number of active calls.  
+- **Using Rate and Parallelism Together**: Both parameters can be combined. For example, with a rate of 10 per second and parallelism of 20, if each request takes a minute to complete, QStash will trigger 10 calls in the first second and another 10 in the next. Since none of them will have finished, the system will wait until one completes before triggering another.  
+
+## How to Use It  
+
+### QStash  
+
+Here’s an example of using Flow-Control with QStash:  
+
+```js
+const client = new Client({ token: "<QStash_TOKEN>" });
+
+await client.publishJSON({
+    url: "https://my-api...",
+    body: { hello: "world" },
+    flowControl: { key: "app1", parallelism: 3, rate: 10 },
+});
+```
+
+For more details and other SDKs, see the documentation [here](https://upstash.com/docs/QStash/features/rate).  
+
+### Workflows  
+
+There are two main use cases for Flow-Control in Workflows:  
+1. [Limiting the Workflow Environment](#limiting-the-workflow-environment): Controlling the execution environment.  
+2. [Limiting External API Calls](#limiting-external-api-calls): Preventing excessive requests to external services.  
+
+#### Limiting the Workflow Environment  
+
+To limit the execution environment, configure the following:  
```

Comment:
Should we provide some details about how it works on step level; across all steps of workflow, this configuration will apply. They might understand it like in step level (e.g. will `step-1` will have parallelism 10?).